File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
    public:
        vector<int> preorderTraversal(TreeNode *root)
        {
            vector<int> treeVal;

            if (!root) return treeVal;

            vector<TreeNode*> tree;
            tree.push_back(root);

            while (!tree.empty())
            {
                auto node = tree.back();
                tree.pop_back();
                treeVal.push_back(node->val);

                if (node->right)
                    tree.push_back(node->right);

                if (node->left)
                    tree.push_back(node->left);
            }

            return treeVal;
        }
};

File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution
{
    public IList<int> PreorderTraversal(TreeNode root)
    {

        List<int> treeVal = new List<int>();

        if (root == null) return treeVal;

        List<TreeNode> tree = new List<TreeNode>();
        tree.Add(root);

        while (tree.Count > 0)
        {
            var node = tree[tree.Count - 1];
            tree.RemoveAt(tree.Count - 1);
            treeVal.Add(node.val);

            if (node.right != null)
                tree.Add(node.right);

            if (node.left != null)
                tree.Add(node.left);
        }

        return treeVal;
    }
}